library/Graphite/Web/Controller/IcingadbGraphiteController.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+namespace Icinga\Module\Graphite\Web\Controller;
+
+use Icinga\Module\Icingadb\Common\Auth;
+use Icinga\Module\Icingadb\Common\Database;
+use Icinga\Module\Icingadb\Common\SearchControls;
+use ipl\Orm\Query;
+use ipl\Orm\UnionQuery;
+use ipl\Stdlib\Contract\Paginatable;
+use ipl\Web\Compat\CompatController;
+use ipl\Web\Control\LimitControl;
+use ipl\Web\Control\PaginationControl;
+use ipl\Web\Control\SortControl;
+use ipl\Web\Filter\QueryString;
+use ipl\Stdlib\Filter;
+use ipl\Web\Url;
+
+class IcingadbGraphiteController extends CompatController
+{
+    use Auth;
+    use Database;
+    use SearchControls;
+
+    /** @var Filter\Rule Filter from query string parameters */
+    private $filter;
+
+    /** @var string|null */
+    private $format;
+
+    /**
+     * Get the filter created from query string parameters
+     *
+     * @return Filter\Rule
+     */
+    public function getFilter(): Filter\Rule
+    {
+        if ($this->filter === null) {
+            $this->filter = QueryString::parse((string) $this->params);
+        }
+
+        return $this->filter;
+    }
+    /**
+     * Create and return the LimitControl
+     *
+     * This automatically shifts the limit URL parameter from {@link $params}.
+     *
+     * @return LimitControl
+     */
+    public function createLimitControl(): LimitControl
+    {
+        $limitControl = new LimitControl(Url::fromRequest());
+        $limitControl->setDefaultLimit($this->getPageSize(null));
+
+        $this->params->shift($limitControl->getLimitParam());
+
+        return $limitControl;
+    }
+
+    public function filter(Query $query, Filter\Rule $filter = null): self
+    {
+        //TODO don't need format?
+        if ($this->format !== 'sql' || $this->hasPermission('config/authentication/roles/show')) {
+            $this->applyRestrictions($query);
+        }
+
+        if ($query instanceof UnionQuery) {
+            foreach ($query->getUnions() as $query) {
+                $query->filter($filter ?: $this->getFilter());
+            }
+        } else {
```

Comment:
You don't feed this with any union query. Remove this.

Reply:
And you don't feed this ever without a filter... I guess it's sufficient to use this in both controllers and to remove this method:

```
$this->applyRestrictions($query);
$query->filter($filter);
```